### PR TITLE
#1121 : ValueSourceAttribute.GetDataSource should throw instead of return null

### PR DESCRIPTION
**Pull request by alfeg**
My attempt to fix bug #1121.

### CODE REVIEW

File: src/NUnitFramework/tests/Attributes/ValueSourceTests.cs [hunk 1]
```diff
@@ -168,6 +172,52 @@ public static IEnumerable<int> IntegerProvider()
...
+            var parameters = method.GetParameters();
+
+            foreach (var parameter in parameters)
+            {
+                var dataSource = parameter.GetCustomAttributes<IParameterDataSource>(false)[0];
+
+                Assert.Throws<InvalidDataSourceException>(() =>
+                {
+                    var data = dataSource.GetData(parameter);
+                    Assert.IsNotNull(data);
```
Comment (rprouse) [inline src/NUnitFramework/tests/Attributes/ValueSourceTests.cs]: I am assuming that the GetData is the line that throws, so is the Assert.IsNotNull necessary?

**Comment by rprouse** [comment]
Very minor comment on my part. This looks good, it just needs a second team member to review.

@CharliePoole, this is easiest to review ignoring whitespace with `?w=1`

**Comment by alfeg** [comment]
I'm not sure that those fix is enough. 

It's still easy to break test suite with non-enumerable values. Like:

``` csharp
public static int NonEnumerable = 1
```

```
1) Invalid : NUnit.Framework.Attributes.ValueSourceTests
An exception was thrown while loading the test.
System.InvalidCastException: Unable to cast object of type 'System.Int32' to type 'System.Collections.IEnumerable'.
   at NUnit.Framework.ValueSourceAttribute.GetDataSourceValue(MemberInfo[] members) in D:\tmp\nunit\src\NUnitFramework\framework\Attributes\ValueSourceAttribute.cs:line 131
   at NUnit.Framework.ValueSourceAttribute.GetDataSource(IParameterInfo parameter) in D:\tmp\nunit\src\NUnitFramework\framework\Attributes\ValueSourceAttribute.cs:line 111
   at NUnit.Framework.ValueSourceAttribute.GetData(IParameterInfo parameter) in D:\tmp\nunit\src\NUnitFramework\framework\Attributes\ValueSourceAttribute.cs:line 93
   at NUnit.Framework.Internal.Builders.ParameterDataSourceProvider.GetDataFor(IParameterInfo parameter) in D:\tmp\nunit\src\NUnitFramework\framework\Internal\Builders\ParameterDat
aSourceProvider.cs:line 67
   at NUnit.Framework.CombiningStrategyAttribute.BuildFrom(IMethodInfo method, Test suite) in D:\tmp\nunit\src\NUnitFramework\framework\Attributes\CombiningStrategyAttribute.cs:lin
e 142
```

Maybe there is also should be a check for object is IEnumerable?

**Comment by rprouse** [comment]
I'm not sure we need to go so far as to check that the source is IEnumerable. If you do, you will just display an error. Personally, I think the error you show above is valid since it is an invalid test and therefore should stop the test run. In issue #1121, we want to carry on with the test run because a null or empty collection could be valid depending on where the data is coming from. The example you give above is just an invalid test that can never work.

**Comment by rprouse** [comment]
This has been out there too long. I am going to merge this before merge conflicts start creeping in. It is a small and well tested change.

Thanks for the contribution.

**Pull request merged**